src/devices/Mpu9250/Mpu9250.cs

```diff
@@ -0,0 +1,983 @@
+            }
+            else
+            {
+                accAverage.X = Math.Abs(accSelfTestAverage.X - accSelfTestAverage.X);
+            }
+
+            if (accFactoryTrim.Y != 0)
+            {
+                accAverage.Y = (accSelfTestAverage.Y - accAverage.Y) / accFactoryTrim.Y;
+            }
+            else
+            {
+                accAverage.Y = Math.Abs(accSelfTestAverage.Y - accSelfTestAverage.Y);
+            }
+
+            if (accFactoryTrim.Z != 0)
+            {
+                accAverage.Z = (accSelfTestAverage.Z - accAverage.Z) / accFactoryTrim.Z;
+            }
+            else
+            {
+                accAverage.Z = Math.Abs(accSelfTestAverage.Z - accSelfTestAverage.Z);
+            }
+
+            return new Tuple<Vector3, Vector3>(gyroAvegage, accAverage);
+        }
+
+        #endregion
+
+        #region I2C
+
+        /// <summary>
+        /// Write data on any of the I2C slave attached to the MPU9250
+        /// </summary>
+        /// <param name="i2cChannel">The slave channel to attached to the I2C device</param>
+        /// <param name="address">The I2C address of the slave I2C element</param>
+        /// <param name="register">The register to write to the slave I2C element</param>
+        /// <param name="data">The byte data to write to the slave I2C element</param>
+        public void I2cWrite(I2cChannel i2cChannel, byte address, byte register, byte data)
+        {
+            // I2C_SLVx_ADDR += 3 * i2cChannel
+            byte slvAddress = (byte)((byte)Register.I2C_SLV0_ADDR + 3 * (byte)i2cChannel);
+            Span<byte> dataout = stackalloc byte[2] { slvAddress, address };
+            _i2cDevice.Write(dataout);
+            // I2C_SLVx_REG = I2C_SLVx_ADDR + 1
+            dataout[0] = (byte)(slvAddress + 1);
+            dataout[1] = register;
+            _i2cDevice.Write(dataout);
+            // I2C_SLVx_D0 =  I2C_SLV0_DO + i2cChannel
+            // Except Channel4
+            byte channelData = i2cChannel != I2cChannel.Slave4 ? (byte)((byte)Register.I2C_SLV0_DO + (byte)i2cChannel) : (byte)Register.I2C_SLV4_DO;
+            dataout[0] = channelData;
+            dataout[1] = data;
+            _i2cDevice.Write(dataout);
+            // I2C_SLVx_CTRL = I2C_SLVx_ADDR + 2
+            dataout[0] = (byte)(slvAddress + 2);
+            dataout[1] = 0x81;
+            _i2cDevice.Write(dataout);
+        }
+
+        /// <summary>
+        /// Read data from any of the I2C slave attached to the MPU9250
+        /// </summary>
+        /// <param name="i2cChannel">The slave channel to attached to the I2C device</param>
+        /// <param name="address">The I2C address of the slave I2C element</param>
+        /// <param name="register">The register to read from the slave I2C element</param>
+        /// <param name="readBytes">The read data</param>
+        public void I2cRead(I2cChannel i2cChannel, byte address, byte register, Span<byte> readBytes)
```

Comment:
How come for writing you only allow to write one byte of data but for reading you allow multiple?

Reply:
It's a device limitation. It's just like that. To write multiple bytes, you have to call the write function multiple times. While for reading, you can read up to 24 bytes.